### PR TITLE
Filter start page events to show only current/upcoming with active Tickets or Talks ( issue : #2237 )

### LINKED ISSUE #2237: Start page: show only current and upcoming events with active Tickets or Talks component
The start page currently lists upcoming events without sufficiently filtering by event relevance or readiness. As a result, events that are not yet actionable for users may appear, even though
• The event has already ended
• Neither ticket sales nor call for speakers is live
• The event is not yet ready for public interaction

This reduces the usefulness and clarity of the start page.

Image

## Expected behaviour

Only events that are both **relevant in time** and **actionable** should appear on the start page.

Specifically, the start page should display
• **Current events** (ongoing)
• **Upcoming events** (starting in the future)

And only if at least one of the following components is live
• **Tickets** (ticket sales live)
• **Talks** (Call for Speakers / CFP live)

## Current behaviour

• Events may appear even if they are past
• Events may appear even if no public component is live
• No clear filtering by component readiness exists

## Proposed solution

• Filter out past events from the start page
• Require at least one active public component
– Tickets component is live
– OR Talks component is live
• Ensure filtering logic is applied consistently across the start page and related listings

## Scope

• Start page event listing logic
• Event time filtering (current and upcoming)
• Component state evaluation (Tickets, Talks)
• No changes to organiser or admin configuration UI required

## Acceptance criteria

• Past events are not shown on the start page
• Events without Tickets or Talks live are not shown
• Events with either Tickets or Talks live are shown
• Behaviour is consistent and predictable
• No regressions in other event listings

## Notes

This improves front page relevance and user experience by ensuring visitors only see events they can actively engage with, either by buying tickets or submitting proposals.

### PR DESCRIPTION
**Pull request by SupRaKoshti**
Closes #2237

## Summary
This PR updates the start page event listing logic to ensure only relevant and actionable events are shown.

## Changes
- Exclude past events
- Include only ongoing and upcoming events
- Require at least one public component to be active:
  - Tickets published OR
  - Talks published

## Result
- Past events are no longer displayed
- Events without active Tickets or Talks are excluded
- Filtering behavior is consistent and predictable

## Summary by Sourcery

Restrict the start page event list to only show current and upcoming events that have at least one active public component (tickets or talks).

Enhancements:
- Filter out past events from the start page event listing.
- Ensure only events with currently active published tickets or published talks appear on the start page.

### CODE REVIEW

File: app/eventyay/presale/views/startpage.py [hunk 1]
```diff
@@ -28,9 +31,33 @@ def get_context_data(self, **kwargs):
...
+                        ) &
+                        (
+                            Q(presale_end__isnull=True) |
+                            Q(presale_end__gte=current_time)
+                        )
+                    )
+                    |
+                    # OR Talks live
+                    Q(talks_published=True)
+                )
```
[the remark this answers is not in the document]
Comment (SupRaKoshti) [reply app/eventyay/presale/views/startpage.py]: Previously, the filter only checked talks_published=True, which could include events with closed CFPs. The updated condition now matches is_cfp_open logic from submission.py, ensuring only actionable events are listed.

File: app/eventyay/presale/views/startpage.py [hunk 2]
```diff
@@ -28,9 +31,33 @@ def get_context_data(self, **kwargs):
...
                 .order_by('date_from')
```
[the remark this answers is not in the document]
Comment (SupRaKoshti) [reply app/eventyay/presale/views/startpage.py]: Thanks for pointing this out! I’ll add targeted unit tests to cover the combinations of past/current/upcoming events, tickets published/presale windows, and talks/CFP status to ensure this filtering logic behaves correctly and prevent regressions.

**Comment by SupRaKoshti** [comment]
One check is failed! I am working on this.

**Comment by MukundC25** [comment]
@SupRaKoshti Thanks for the PR, please share screenshots/video of the changes and update the PR title correctly.

**Comment by SupRaKoshti** [comment]
Hi @MukundC25 , Thanks for the review! 

**What I'm working on:**
- ✅ Adding comprehensive unit tests for all filtering scenarios (addressing @copilot feedback)
- 🔄 Fixing the CI test failure (cfp__is_open cannot be filtered as it's a @cached_property)
- 📸 Preparing screenshots/video demonstrating the functionality

**Test scenarios being covered:**
1. ✅ Past events are correctly hidden
2. ✅ Events with active ticket presale windows are shown
3. ✅ Events with closed presale are hidden (if no open CFP)
4. ✅ Events with open CFP are shown
5. ✅ Events with closed CFP are hidden (if no active tickets)

The CI failure is because `is_open` is a cached property, not a database field. I'm replicating its logic using actual database field queries.

Screenshots and test implementation coming shortly!

**Comment by SupRaKoshti** [comment]
Hi @MukundC25 and @mariobehling  👋

I've encountered a merge conflict with PR #2254 that's already been merged into develop. Both PRs implement logic for which events show on the startpage, but with **fundamentally different philosophies**.

**The conflict:**

**My PR implements:** A fully automatic, hard-coded query using `visible_on_startpage()` that dynamically filters events based on:
- Ticket availability (tickets published + presale window)
- CfP availability (talks published + CfP deadline)
- Event is live and not past date_to
- ⚠️ **No admin interaction required** - events appear automatically if they meet criteria

**PR #2254 implements:** Admin-controlled visibility using manual flags:
- `featured_events` - admin sets `startpage_featured=True` flag
- `upcoming_events` - admin sets `startpage_visible=True` flag  
- `past_events` - events that have already ended
- ✅ **Full admin control** - organizers decide what appears on startpage

**Key philosophical difference:**

| Aspect | My Implementation (Automatic) | PR #2254 (Manual Control) |
|--------|-------------------------------|---------------------------|
| **Admin interaction** | ❌ None required - fully automatic | ✅ Admin explicitly chooses visibility |
| **Event selection** | Dynamic based on ticket/CfP availability | Manual flags (`startpage_featured`, `startpage_visible`) |
| **Categorization** | Single list, automatic relevance | Three explicit categories (featured/upcoming/past) |
| **Past events** | ❌ Hidden if no active tickets/CfP | ✅ Can be shown if admin wants |
| **Use case** | "Show all relevant events automatically" | "Let organizers curate what's visible" |
| **Flexibility** | Lower - rules are hard-coded | Higher - admins have full control |

**The fundamental question:**

Should the startpage be:
1. **Fully automatic** (my approach) - Events appear based on hard-coded business rules
2. **Admin-curated** (PR #2254) - Organizers explicitly control what's featured/visible

**My understanding now:**

Looking at PR #2254, it seems the project wants **admin control** over startpage visibility, which makes sense for:
- Allowing organizers to curate their presence
- Featuring special/important events manually
- Showing past events for historical reference
- Giving flexibility beyond automatic rules

**Proposed resolution:**

Since PR #2254 is already merged and provides the admin control the project likely wants, I suggest:

**Option A (Recommended):** Close my PR in favor of PR #2254's approach
- The manual flag system gives more flexibility
- Admins can control exactly what appears
- My automatic logic might be too rigid for real-world needs

**Option B:** Enhance PR #2254 with optional automatic filtering
- Keep the flag-based system as primary
- Add my dynamic logic as a fallback for events without flags set
- But this adds complexity that may not be needed

**Option C:** Keep my approach
- Only if the project wants fully automatic startpage population
- But this removes admin control entirely

I'm leaning toward **Option A** since admin control seems to be the better design pattern here. What do you think?

Happy to close this PR or rebase/refactor based on your guidance! 🙂

Thanks!

**Comment by mariobehling** [comment]
There is no conflict. The different features are described in different issues and should be implemented accordingly.

**Comment by SupRaKoshti** [comment]
Hi @mariobehling ! Since PR #2254 is merged with manual startpage_visible flags, should my automatic filtering in #2244:

Work on top of the flags (events need BOTH admin approval AND active tickets/CFP)?
Replace the manual system entirely with automatic logic?
Be a fallback (use automatic logic only for events without flags set)?

The querysets currently conflict - his uses featured_events/upcoming_events/past_events with manual flags, mine uses a single events list with automatic filtering. Which approach should I follow?

**Comment by SupRaKoshti** [comment]
Hi @mariobehling @MukundC25,

I've implemented the complete solution for issue #2237 with proper queryset method, view integration, and comprehensive test coverage.

## 📋 Implementation Summary:

### **1. New Queryset Method** (`models.py`):
Created `visible_on_startpage()` that filters events based on:
- ✅ **Time relevance**: Only current/upcoming events (using `date_to__date__gte=today`)
- ✅ **Active Tickets component**: `tickets_published=True` AND within presale window
- ✅ **Active Talks/CFP component**: `talks_published=True` AND CFP deadline not passed
- ✅ **Proper joins**: Includes `select_related("organizer", "cfp")` for performance
- ✅ **Distinct results**: Prevents duplicates from submission type joins

### **2. Updated View** (`startpage.py`):
```python
ctx['featured_events'] = qs.filter(startpage_featured=True, testmode=False)
ctx['upcoming_events'] = qs.filter(startpage_visible=True).filter(...)
# No past_events - per issue #2237 requirement
```
Integrates the business logic with admin visibility controls while excluding past events entirely.

### **3. Comprehensive Test Coverage** (19 tests):

**✅ Tickets Component Tests:**
- Active ticket sales (within presale window) → visible
- Presale not started yet → not visible
- Presale expired → not visible
- Null presale dates → visible (no restrictions)

**✅ Talks/CFP Component Tests:**
- Open CFP deadline → visible
- Open submission type deadline → visible
- Null CFP deadline (always open) → visible
- Expired CFP → not visible
- No CFP object → not visible

**❌ Exclusion Tests:**
- Past events with active tickets → not visible (issue #2237)
- Past events with open CFP → not visible (issue #2237)
- Events not live → not visible
- Events without active components → not visible

**✅ Edge Cases:**
- Events ending today → visible
- Events with both components active → visible
- Distinct results (no duplicates) → verified

## 🎯 What This Solves:

✅ **Only current and upcoming events appear** - past events completely excluded  
✅ **Only actionable events appear** - requires active ticket sales OR open CFP  
✅ **Admin controls respected** - integrates with `startpage_visible` and `startpage_featured` flags  
✅ **Performance optimized** - proper `select_related()` and `distinct()` usage  
✅ **Well tested** - 19 tests covering all scenarios  

## 📝 Per Issue #2237 Requirements:

> "Only events that are both **relevant in time** and **actionable** should appear on the start page."

✅ **Relevant in time**: Current or upcoming events only (past events excluded)  
✅ **Actionable**: At least one public component must be active (Tickets OR Talks)  

## ⚠️ Note on Past Events:

Issue #2237 explicitly states: *"Past events are not shown on the start page"* - therefore I have **not** included `past_events` in the context. If you need past events displayed in a separate section, please let me know and I'll add that functionality.

All tests are passing. Please review and let me know if any adjustments are needed!

---

**Files Changed:**
- `app/eventyay/base/models/event.py` - Added `visible_on_startpage()` queryset method
- `app/eventyay/presale/views/startpage.py` - Updated view to use new queryset
- `app/tests/talk/event/test_event_queryset.py` - Added 19 comprehensive tests

Screenshot 2026-02-07 215950
Screenshot 2026-02-07 220018

File: app/tests/talk/event/test_event_queryset.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
+from django.test import TestCase
+from django.utils.timezone import now, timedelta
+
+from eventyay.base.models.event import Event
+from eventyay.base.models.organizer import Organizer
+from eventyay.base.models.cfp import CfP
```
[the remark this answers is not in the document]
Comment (SupRaKoshti) [reply app/tests/talk/event/test_event_queryset.py]: I have imported timedelta from django.utils.timezone but it doesn't give me any error event though i will import the timedelta from the datetime!

Yes, Cfp is unused import here so i will also remove that!

File: app/tests/talk/event/test_event_queryset.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+    # ===============================
+    # ✅ Edge Cases
+    # ===============================
+
+    def test_event_ending_today_is_visible(self):
+        """Event ending today should still be visible"""
+        from django.utils import timezone
+        
+        today = timezone.localdate()
```
[the remark this answers is not in the document]
Comment (SupRaKoshti) [reply app/tests/talk/event/test_event_queryset.py:333]: okay, I will moved the import in module import section.

**Comment by SupRaKoshti** [comment]
Hi @Saksham-Sirohi 👋

I've addressed all the review comments and pushed the updates.
All CI checks are passing now.

Please let me know if any further changes are needed. 🙂

**Comment by SupRaKoshti** [comment]
Hi @mariobehling :),
I noticed similar changes were merged in PR #2370 .
Should I close this PR to avoid duplication?

**Comment by mariobehling** [comment]
Closing this PR as it has been addressed already.

Also a small process note.

We have automatic Copilot PR reviews enabled on this repository. These reviews are only triggered if the contributor has GitHub Copilot enabled and an active license on their own account.

Please enable Copilot in your GitHub settings if you have access. In many regions, free licenses are available through educational institutions or developer programs. Enabling Copilot helps us speed up the auto review process and reduces manual review overhead for the core team.

Thank you for your contribution and cooperation.
